### ISSUE TITLE
Relatively slow processing

### ISSUE THREAD

**Issue by vadik299** [UglyToad/PdfPig#47]
This is not exactly an issue, but more like a general question. While extracting Letters collection I noticed that overall the process runs about 4-5 times slower than PdfBox. I run PdfBox through Ikvm, so I was expecting it to be the other way around :-)
Of course there can be many things contributing to this, but I did one quick test - I ran a mass replace of word "decimal" to "double" across the whole code base. And yes, the speed got right on par with PdfBox! Changing it to "float" made it even a little faster (probably due to smaller memory footprint).
Sure double/float is not precise, but personally I often need to run extraction over hundreds of thousands  PDFs, so speed is crucial and the time difference is substantial. On the other hand I think that letters/lines positions and dimensions would be OK with 2 digits precision at most (ok, maybe 3:-))
I see a few ways to approach this:
1. Change all properties to float. This is quick, but not very user friendly and may have some minor issues if anyone ever tries to compare numbers directly.
2. Keep values internally as Int multiplied by 100 (or 1000). Do all calculations on Int, then return to the user as decimal (divide by 100) - this may be better, but probably harder to do and potentially more confusing.
3. Keep everything as is and suggest getting a better server :-) (or optimize somewhere else)

Thoughts? Thank you for your time and all the work put in this library!

**Comment by EliotJones** [comment]
Thanks for the feedback, it's useful to know the performance characteristics in production. 

I'm surprised it's so much slower! I'm going to run the profiling this weekend but it sounds like the decimals are a large part of it.

The reason for using decimals over doubles initially is that it allows accurate representation of the numbers as they appear in the document, so for example a font size is reported as 12.5 rather than 12.499999 or whatever else doubles get up to.

It seems like where multiple operations are being done on those decimal values though (matrix transformations for processing a page content stream) they could get away with being doubles or even floats then cast back to decimals on the public API. This keeps decimals on the public facing API keeping the values true to what was reported on the face of the document for things like writing or reading numeric values, but speeds up processing which loses some accuracy anyway.

I'll look at the feasibility of replacing decimals during calculations and the corresponding speed increase over the weekend. I've got some tests which confirm the calculated values are within some tolerance of the PDFBox values so as long as they still pass and the speed increases I can't see any problems https://github.com/UglyToad/PdfPig/blob/master/src/UglyToad.PdfPig.Tests/Integration/SinglePageSimpleGoogleChromeTests.cs#L107. I'd hoped that it would be fast enough without having to do that but performance was one of the aims of the library so it sounds like it's falling short of that. One thing that hopefully is supported is parallel processing of documents but I've not tested that yet either.

(btw I haven't forgotten your PR it's just been a super busy week, sorry!)

**Comment by vadik299** [comment]
Thank you Eliot! Appreciate the answer. If you end up using floats internally I'd suggest to have ability to access them directly as well (e.g. through ExperimentalAccess) since float->decimal conversion may take up some time too.

**Comment by BobLd** [comment]
Agreed with vadik299 on the use of `float` over `decimal` and `double`. The `float` accuracy should be more than enough for this use.

**Comment by tincann** [comment]
I've also noticed these performance characteristics as described by @vadik299. I've tested it locally on a batch of documents and compared the timing with another library. It indeed showed a 5 times difference. The reason for this test is because we suddenly noticed big slowdowns in our pdf processing webjob after switching to PdfPig. For example, one batch took 10 minutes, but strangely when tested locally it only takes 5 seconds. To be fair, I haven't done any remote profiling yet, so the problem might lie somewhere else.

Albeit hacky, I'm tempted to go with vadik's solution (as a temporary workaround) to mass replace decimals with floats to see if this solves the problem in our web apps. However, I'm not sure whether there are some decimals that actually need to stay decimals.

On the whole, I really like the library. Next to that it has a sensible API, it's also the only library for .net core that I have found that correctly decodes cmap fonts for our use case.

**Comment by Numpsy** [comment]
Is it worth creating a BenchmarkDotNet project to get some figures on the performance? (useful for testing the effect of any changes).

**Comment by EliotJones** [comment]
Thanks for the additional information on this and sorry it's taking so long to get anything done on this (I'm not going to have much time for substantive changes until I leave my current job at the end of the month).

@tincann this is interesting because it matches something I was seeing on Appveyor, CI builds were routinely timing out on the tests despite the timeout being set to 1 hr 30 mins. This resulted in the switch to Azure Pipelines which now runs in about 6 minutes (except now because I just broke it 😆). I think there's some difference specific to the machine which causes a huge performance hit, my initial guess is it's IO bound. When I run locally I'm usually running off an SSD and I'd assume the free tier AppVeyor boxes are slow HDDs, but it's impossible to diagnose because I don't have access to the actual box.

I think as you say it's worth trying the float approach, I can try that on a separate branch and then we could see what performance improvement it delivers, it's definitely part of the performance issues people are encountering, but I'm not sure to what extent it is vs. something else.

It's also definitely a good idea to add Benchmarking as @Numpsy mentions, I'll look into doing that as part of the performance work.

I've started running some dotTrace profiling this morning just to get more of an idea where time is going. This is the profile of a console application which loads the 26 documents from the integration tests and reads every page summing the count of letters:

![image](https://user-images.githubusercontent.com/7862346/63224152-114e5400-c1b8-11e9-8f62-1c4de2905f1b.png)

This takes about 10.4 seconds on my machine, which is sub-optimal but probably within my tolerance for overall performance (one of the trade-offs in the library is exposing individual letter information by default vs just the text data as in PDFBox).

Of that about 3 seconds is spent in TransformationMatrix code so the switch to doubles should have a large impact here, but that still leaves ~70% of the time elsewhere. I need to go slightly more in depth before stating anything for certain and definitely need BenchmarkDotNet coverage as @Numpsy suggests; but I'm hopeful/wondering if there's something else causing this perf killing slowdown people are encountering.

**Comment by tincann** [comment]
@EliotJones Thanks for the elaborate answer. I've checked the processing code on our side, and it appears that the library gets called through the `PdfDocument.Open(Stream stream)` API. The stream that's passed in comes from `IFormFile.OpenReadStream()`, so from a HTTP POST action (called by a web job). First intuition says that this might be the culprit as it is a network stream. However, before PdfPig tries to read from this stream, it first gets fully read due to it being saved to the blob storage, then seeked (sought?) to the beginning, and then finally passed to PdfPig. 

Not entirely sure, but can I safely conclude that this means that PdfPig is simply reading a memory stream because of the above? If that's true, then the performance difference between local and web app can not be explained by an IO bottleneck (using a network stream or HDD vs SSD). Please correct me if I'm wrong.

PS: You don't have to apologize for things taking long. I think everyone here appreciates the hard work you're putting into this in your free time!

**Comment by Numpsy** [comment]
My experience with file handling in ASP.Net Core is rather limited, but from what i recall, the file streams you get from it can be either memory based or file based depending on circumstances (e.g. files above a certain size might be buffered to disk wheras small files might not), which might effect things? (can't say anything about other web frameworks).

**Comment by tincann** [comment]
@Numpsy I have solved the problem by first copying the network stream to a memory stream and then feeding it to PdfPig. So the culprit was the network stream after all. I guess the previous pdf library that we used already did that behind the scenes (or it doesn't use the stream in the same way as PdfPig).

**Comment by EliotJones** [comment]
I'm glad to hear passing the data into a `MemoryStream` improves the performance. PdfPig consumes the stream in a fairly naive way by just seeking/reading from it directly which is incredibly bad in almost every stream use-case except for MemoryStream. 

I think @Numpsy is right that a network based stream like that is going to be buffered somewhere by the framework/OS where reading from it is probably not going to be performant.

A better optimization might be for PdfPig to wrap the input stream as a `BufferedStream` which would work around the terrible performance of many streams when reading repeatedly - https://docs.microsoft.com/en-us/dotnet/api/system.io.bufferedstream?view=netframework-4.8. One of the main performance hits from profiling seems to be repeated reads from both Stream and Byte Array inputs, there's probably a benefit to adding some additional intelligent buffering behaviour inside PdfPig.

I've found a few low-hanging fruit and managed to shave about 4-5 seconds off the performance of my little benchmark when running in Tracing mode in dotTrace but there's a lot more to be done.

**Comment by vadik299** [comment]
Guys, thanks for all the optimization efforts!

I put together a small benchmark comparing a few versions of PdfPig:
- release 0.9
- release 0.9 with mass-replaced decimal->float
- latest commits with optimizations done by Eliot (great job there!)
- same one, with mass-replaced decimal->float again
- PdfBox 1.8

I ran a few files with each version 3 times, and below are the log results. It seems that changing decimal to float makes it on average two times faster, but it varies with different files. 

Also for some files I got different letter count between PdfPig and PdfBox. Not quite sure why it's happening, will try to look more into it.

I'm attaching the benchmark project here, since I'm not sure where to put it. (it uses a mix of DLLs, so doesn't look like it belongs to the main project)

[PdfPigBenchmark.zip](https://github.com/UglyToad/PdfPig/files/3553587/PdfPigBenchmark.zip)


-------------------------------------------------------------------------
Starting benchmark. Reading each file 3 times

Running file Sample00_5_pages.pdf

PdfPig 0.9               ... Time (ms): 446  Letters: 7402
PdfPig 0.9-FLOAT         ... Time (ms): 304  Letters: 7402
PdfPig-Optimized         ... Time (ms): 201  Letters: 7402
PdfPig-Optimized-FLOAT   ... Time (ms): 146  Letters: 7402
PdfBox 1.8               ... Time (ms): 321  Letters: 7402

Running file Sample05_29_pages.pdf

PdfPig 0.9               ... Time (ms): 1722  Letters: 76443
PdfPig 0.9-FLOAT         ... Time (ms): 701  Letters: 76443
PdfPig-Optimized         ... Time (ms): 989  Letters: 76443
PdfPig-Optimized-FLOAT   ... Time (ms): 414  Letters: 76443
PdfBox 1.8               ... Time (ms): 1250  Letters: 76363

Running file Sample10_64_pages.pdf

PdfPig 0.9               ... Time (ms): 6447  Letters: 103719
PdfPig 0.9-FLOAT         ... Time (ms): 3832  Letters: 103719
PdfPig-Optimized         ... Time (ms): 4556  Letters: 103719
PdfPig-Optimized-FLOAT   ... Time (ms): 2627  Letters: 103719
PdfBox 1.8               ... Time (ms): 3243  Letters: 103719

Running file Sample20_397_Pages.pdf

PdfPig 0.9               ... Time (ms): 36394  Letters: 1347928
PdfPig 0.9-FLOAT         ... Time (ms): 16549  Letters: 1347928
PdfPig-Optimized         ... Time (ms): 23342  Letters: 1347928
PdfPig-Optimized-FLOAT   ... Time (ms): 11465  Letters: 1347928
PdfBox 1.8               ... Time (ms): 16514  Letters: 1349279

Running file Sample30_3024_pages.pdf

PdfPig 0.9               ... Time (ms): 8088210  Letters: 14463498
PdfPig 0.9-FLOAT         ... Time (ms): 260973  Letters: 14463498
PdfPig-Optimized         ... Time (ms): 340024  Letters: 14463498
PdfPig-Optimized-FLOAT   ... Time (ms): 188554  Letters: 14463498
PdfBox 1.8               ... Time (ms): 40569  Letters: 14463498

Benchmark completed. Press ENTER to close

**Comment by tincann** [comment]
@vadik299 Wow! That looks very promising. Nice work!

**Comment by vadik299** [comment]
@tincann It's a good point! I added count of words and in one document it is slightly different. 
Not sure how bad it really is.


----
Starting benchmark. Reading each file 3 times

Running file Sample00_5_pages.pdf

PdfPig 0.9               ... Time (ms): 431  Letters: 7402  Words: 1619
PdfPig 0.9-FLOAT         ... Time (ms): 287  Letters: 7402  Words: 1619
PdfPig-Optimized         ... Time (ms): 208  Letters: 7402  Words: 1619
PdfPig-Optimized-FLOAT   ... Time (ms): 146  Letters: 7402  Words: 1619
PdfBox 1.8               ... Time (ms): 331  Letters: 7402

Running file Sample05_29_pages.pdf

PdfPig 0.9               ... Time (ms): 1745  Letters: 76443  Words: 18790
PdfPig 0.9-FLOAT         ... Time (ms): 741  Letters: 76443  Words: 18790
PdfPig-Optimized         ... Time (ms): 1076  Letters: 76443  Words: 18790
PdfPig-Optimized-FLOAT   ... Time (ms): 417  Letters: 76443  Words: 18790
PdfBox 1.8               ... Time (ms): 1107  Letters: 76363

Running file Sample10_64_pages.pdf

PdfPig 0.9               ... Time (ms): 6288  Letters: 103719  Words: 20113
PdfPig 0.9-FLOAT         ... Time (ms): 3756  Letters: 103719  Words: 20113
PdfPig-Optimized         ... Time (ms): 4433  Letters: 103719  Words: 20113
PdfPig-Optimized-FLOAT   ... Time (ms): 2636  Letters: 103719  Words: 20113
PdfBox 1.8               ... Time (ms): 2916  Letters: 103719

Running file Sample20_397_Pages.pdf

PdfPig 0.9               ... Time (ms): 36811  Letters: 1347928  Words: **319997**
PdfPig 0.9-FLOAT         ... Time (ms): 16365  Letters: 1347928  Words: **320003**
PdfPig-Optimized         ... Time (ms): 23421  Letters: 1347928  Words: 319997
PdfPig-Optimized-FLOAT   ... Time (ms): 10872  Letters: 1347928  Words: 320003
PdfBox 1.8               ... Time (ms): 15095  Letters: 1349279

Running file Sample30_3024_pages.pdf

PdfPig 0.9               ... Time (ms): 297589  Letters: 14463498  Words: 273865
PdfPig 0.9-FLOAT         ... Time (ms): 134848  Letters: 14463498  Words: 273865
PdfPig-Optimized         ... Time (ms): 205421  Letters: 14463498  Words: 273865
PdfPig-Optimized-FLOAT   ... Time (ms): 107753  Letters: 14463498  Words: 273865
PdfBox 1.8               ... Time (ms): 26653  Letters: 14463498

Benchmark completed. Press ENTER to close

**Comment by EliotJones** [comment]
Thank you for this very useful analysis, it's certainly going to help with optimization. It looks like the performance is almost acceptable until after the 397 page case. I'm going to start rolling floats up into the calculation code from the places it makes sense to do so (TrueType defines things using floats rather than decimals, for example) over time.

I'll also have a more in depth look into what causes the huge time lag in the 3024 case, I'd imagine at that point it's just the number of calculations dominating, the time suggests this is the case, I think decimals are around 20 times slower than floats. However it's concerning that even the optimized with floats version is so much slower than PDFBox.

I'm not too worried about discrepancies in word and letter counts as long as the coordinates are within tolerance of the values reported by PDFBox but it'll be worth looking into, I'm just not sure I'll find the motivation to do so 😆

**Comment by vadik299** [comment]
I ran benchmark for the 0.1.0 release. The speed has greatly improved! 
(2-4 times comparing to 0.0.9 and 1.5-2 times comparing to "optimized" version that was made available around end of August. Wow!!!)

But still, "float" version is much faster :-p

A simple way to make *everything* float is this: do search and replace over the whole solution for:
1. "decimal" to "float" (case-sensitive replace)
2. "Decimal" to "Float" (case-sensitive replace)
3. "(?<=[0-9])m" to "f" (regex replace) - this replaces numerical constants
Then fix a few minor errors and that's it!

Below are the benchmark results. Please note that letter count for some documents decreased. Not sure if that represents a problem or not.

Starting benchmark. Reading each file 3 times

Running file Sample00_5_pages.pdf

PdfPig 0.9               ... Time (ms): 365  Letters: 7402
PdfPig 0.9-FLOAT         ... Time (ms): 365  Letters: 7402
PdfPig-Optimized         ... Time (ms): 211  Letters: 7402
PdfPig-Optimized-FLOAT   ... Time (ms): 147  Letters: 7402
PdfPig 0.10              ... Time (ms): 212  Letters: 7402
PdfPig 0.10float         ... Time (ms): 168  Letters: 7402
PdfBox 1.8               ... Time (ms): 363  Letters: 7402

Running file Sample05_29_pages.pdf

PdfPig 0.9               ... Time (ms): 1847  Letters: 76443
PdfPig 0.9-FLOAT         ... Time (ms): 827  Letters: 76443
PdfPig-Optimized         ... Time (ms): 1123  Letters: 76443
PdfPig-Optimized-FLOAT   ... Time (ms): 539  Letters: 76443
PdfPig 0.10              ... Time (ms): 967  Letters: 76363
PdfPig 0.10float         ... Time (ms): 378  Letters: 76363
PdfBox 1.8               ... Time (ms): 1349  Letters: 76363

Running file Sample10_64_pages.pdf

PdfPig 0.9               ... Time (ms): 7848  Letters: 103719
PdfPig 0.9-FLOAT         ... Time (ms): 5084  Letters: 103719
PdfPig-Optimized         ... Time (ms): 6385  Letters: 103719
PdfPig-Optimized-FLOAT   ... Time (ms): 4004  Letters: 103719
PdfPig 0.10              ... Time (ms): 4459  Letters: 103719
PdfPig 0.10float         ... Time (ms): 2579  Letters: 103719
PdfBox 1.8               ... Time (ms): 3732  Letters: 103719

Running file Sample20_397_Pages.pdf

PdfPig 0.9               ... Time (ms): 45746  Letters: 1347928
PdfPig 0.9-FLOAT         ... Time (ms): 22507  Letters: 1347928
PdfPig-Optimized         ... Time (ms): 30956  Letters: 1347928
PdfPig-Optimized-FLOAT   ... Time (ms): 15808  Letters: 1347928
PdfPig 0.10              ... Time (ms): 21728  Letters: 1349279
PdfPig 0.10float         ... Time (ms): 8710  Letters: 1349279
PdfBox 1.8               ... Time (ms): 22669  Letters: 1349279

Running file Sample30_3024_pages.pdf

PdfPig 0.9               ... Time (ms): 387481  Letters: 14463498
PdfPig 0.9-FLOAT         ... Time (ms): 197769  Letters: 14463498
PdfPig-Optimized         ... Time (ms): 276626  Letters: 14463498
PdfPig-Optimized-FLOAT   ... Time (ms): 155107  Letters: 14463498
PdfPig 0.10              ... Time (ms): 96754  Letters: 14463498
PdfPig 0.10float         ... Time (ms): 17949  Letters: 14463498
PdfBox 1.8               ... Time (ms): 31665  Letters: 14463498

Benchmark completed. Press ENTER to close

**Comment by EliotJones** [comment]
Hi @vadik299, I have finally decided its worth doing this for the next release since I was testing with a few hundred documents and it's very slow at the moment.

I was wondering about using double instead of float since on most 64 bit processors doubles have native support. Would you be able to profile float against double please to check the timings are within a few milliseconds of each other. That way I won't have to go through and replace everything with floats after doing doubles.

Thanks

**Comment by vadik299** [comment]
Hi @EliotJones, below are the benchmark results. I'm not sure how to interpret them though :-)

I ran it a few times and it was jumping back-and-fourth a little. Sometimes float runs were a bit faster, sometimes double. So like you expected, it's within a few milliseconds from each other. I have 64 bit OS, Intel i7 CPU. 

Personally, I think if performance is the same, float should be better, due to smaller memory consumption. It also seems to be reflected in the benchmark - larger files seem slower with double, probably because of more time required for memory allocation. In terms of precision float should be sufficient for pretty much everything within PDF. One advantage of double may be that it requires less casting to use in Math functions, but that's minimal and probably not always applicable.
Let me know if there's anything else you'd like me to test. Will be happy to run it.

Starting benchmark. Reading each file 6 times

Running file Sample00_5_pages.pdf

PdfPig 0.10float         ...... Time (ms): 174  Letters: 7402
PdfPig 0.10double        ...... Time (ms): 167  Letters: 7402

Running file Sample05_29_pages.pdf

PdfPig 0.10float         ...... Time (ms): 778  Letters: 76363
PdfPig 0.10double        ...... Time (ms): 644  Letters: 76363

Running file Sample10_64_pages.pdf

PdfPig 0.10float         ...... Time (ms): 4265  Letters: 103719
PdfPig 0.10double        ...... Time (ms): 4399  Letters: 103719

Running file Sample20_397_Pages.pdf

PdfPig 0.10float         ...... Time (ms): 14784  Letters: 1349279
PdfPig 0.10double        ...... Time (ms): 15532  Letters: 1349279

Running file Sample30_3024_pages.pdf

PdfPig 0.10float         ...... Time (ms): 32270  Letters: 14463498
PdfPig 0.10double        ...... Time (ms): 33693  Letters: 14463498

Benchmark completed. Press ENTER to close

**Comment by EliotJones** [comment]
Thank you very much for running those. They're about in-line with what I expected. As mentioned in the pull request #106 I'm going with doubles for now because, as you mention, it's easier for the user to pass those to math functions, etc and more standard for an API.

I plan however to add additional options to `ParsingOptions` which would prevent many of the things that current get added to memory (and then GCed) being created in the first place. This should support high-performance workflows where certain details aren't needed on the public API.

**Comment by EliotJones** [comment]
Hi @vadik299 , if you get time would you be able to run the attached package against your benchmark code please? Maybe compare current 0.0.11 against mass float replace and then the attached package. The attached package is the code from the PR so I want to check the places I avoided replacing decimals aren't causing a huge slowdown 🤞 . Thanks :)

[PdfPig.0.0.11.1.zip](https://github.com/UglyToad/PdfPig/files/3992657/PdfPig.0.0.11.1.zip)

**Comment by vadik299** [comment]
Hi @EliotJones, sure! The results are below. Seems like the new code is much faster! 3 files out of 5 got better results than float/double version.

Version 0.10 in the benchmark below is actually 0.0.11 (I downloaded it right before 0.0.11 was released, but it already included most of the commits). And 0.11 is the version that you provided.
This benchmark extracts all letters and all paths.

Starting benchmark. Reading each file 6 times

Running file Sample00_5_pages.pdf

PdfPig 0.10              ...... Time (ms): 264  Letters: 7402  Paths: 265
PdfPig 0.10float         ...... Time (ms): 188  Letters: 7402  Paths: 265
PdfPig 0.10double        ...... Time (ms): 162  Letters: 7402  Paths: 265
PdfPig 0.11              ...... Time (ms): 175  Letters: 7402  Paths: 265

Running file Sample05_29_pages.pdf

PdfPig 0.10              ...... Time (ms): 1578  Letters: 76363  Paths: 12801
PdfPig 0.10float         ...... Time (ms): 846  Letters: 76363  Paths: 12801
PdfPig 0.10double        ...... Time (ms): 741  Letters: 76363  Paths: 12801
PdfPig 0.11              ...... Time (ms): 684  Letters: 76363  Paths: 12801

Running file Sample10_64_pages.pdf

PdfPig 0.10              ...... Time (ms): 7023  Letters: 103719  Paths: 203129
PdfPig 0.10float         ...... Time (ms): 4155  Letters: 103719  Paths: 203129
PdfPig 0.10double        ...... Time (ms): 3997  Letters: 103719  Paths: 203129
PdfPig 0.11              ...... Time (ms): 5314  Letters: 103719  Paths: 203129

Running file Sample20_397_Pages.pdf

PdfPig 0.10              ...... Time (ms): 32053  Letters: 1349279  Paths: 345120
PdfPig 0.10float         ...... Time (ms): 13221  Letters: 1349279  Paths: 345120
PdfPig 0.10double        ...... Time (ms): 12772  Letters: 1349279  Paths: 345120
PdfPig 0.11              ...... Time (ms): 15047  Letters: 1349279  Paths: 345120

Running file Sample30_3024_pages.pdf

PdfPig 0.10              ...... Time (ms): 140692  Letters: 14463498  Paths: 190453
PdfPig 0.10float         ...... Time (ms): 26027  Letters: 14463498  Paths: 190453
PdfPig 0.10double        ...... Time (ms): 20099  Letters: 14463498  Paths: 190453
PdfPig 0.11              ...... Time (ms): 18912  Letters: 14463498  Paths: 190453

Benchmark completed. Press ENTER to close

**Comment by EliotJones** [comment]
Thanks very much, I think I should be ready to merge the change to double in now. I ran another set of benchmarks with the released 0.0.11, PDFBox, the conversion to doubles and the conversion to doubles plus other optimizations, these were the results:

```
Starting benchmark. Reading each file 3 times

Running file Sample00_5_pages.pdf

PdfPig 0.11              ... Time (ms): 228  Letters: 7402
PdfPig-Optimized         ... Time (ms): 186  Letters: 7402
PdfBox 1.8               ... Time (ms): 350  Letters: 7402
PdfPig Doubles           ... Time (ms): 197  Letters: 7402

Running file Sample05_29_pages.pdf

PdfPig 0.11              ... Time (ms): 954  Letters: 76363
PdfPig-Optimized         ... Time (ms): 305  Letters: 76363
PdfBox 1.8               ... Time (ms): 1215  Letters: 76363
PdfPig Doubles           ... Time (ms): 346  Letters: 76363

Running file Sample10_64_pages.pdf

PdfPig 0.11              ... Time (ms): 3412  Letters: 103719
PdfPig-Optimized         ... Time (ms): 2014  Letters: 103719
PdfBox 1.8               ... Time (ms): 3399  Letters: 103719
PdfPig Doubles           ... Time (ms): 2476  Letters: 103719

Running file Sample20_397_Pages.pdf

PdfPig 0.11              ... Time (ms): 16191  Letters: 1349279
PdfPig-Optimized         ... Time (ms): 6335  Letters: 1349279
PdfBox 1.8               ... Time (ms): 16805  Letters: 1349279
PdfPig Doubles           ... Time (ms): 6894  Letters: 1349279

Running file Sample30_3024_pages.pdf

PdfPig 0.11              ... Time (ms): 78539  Letters: 14463498
PdfPig-Optimized         ... Time (ms): 9142  Letters: 14463498
PdfBox 1.8               ... Time (ms): 24288  Letters: 14463498
PdfPig Doubles           ... Time (ms): 9000  Letters: 14463498

Benchmark completed. Press ENTER to close
```

Annoyingly the additional optimizations for some reason made the 3024 case slower, but they should reduce memory usage in all cases.

**Comment by vadik299** [comment]
Looks like this can be closed now. Performance is great now, thank you for all the improvements done!
